include/nbl/system/IFileArchive.h

```diff
@@ -1,121 +1,115 @@
+enum EFileSystemType
+{
+	FILESYSTEM_NATIVE = 0,	// Native OS FileSystem
+	FILESYSTEM_VIRTUAL	// Virtual FileSystem
+};
+
+// @sadiuk Move this enum into IFileArchive
+//! Contains the different types of archives
+enum E_FILE_ARCHIVE_TYPE
+{
+	//! A PKZIP archive
+	EFAT_ZIP,
+
+	//! A gzip archive
+	EFAT_GZIP,
+
+	//! A virtual directory
+	EFAT_FOLDER,
+
+	//! An ID Software PAK archive
+	EFAT_PAK,
+
+	//! A Tape ARchive
+	EFAT_TAR,
+
+	//! The type of this archive is unknown
+	EFAT_UNKNOWN
+};
+
+//! The FileArchive manages archives and provides access to files inside them.
+class IFileArchive : public virtual core::IReferenceCounted
+{
+public:
+	struct SOpenFileParams
+	{
+		std::string password;
+	};
+
+	//! Opens a file based on its name
+	/** Creates and returns a new IReadFile for a file in the archive.
+	\param filename The file to open
+	\return Returns A pointer to the created file on success,
+	or 0 on failure. */
+	virtual core::smart_refctd_ptr<IReadFile> createAndOpenFile(const std::filesystem::path& filename, const SOpenFileParams& params) = 0;
+
+	//! Returns the complete file tree
+	/** \return Returns the complete directory tree for the archive,
+	including all files and folders */
+	virtual const IFileList* getFileList() const = 0;
+
+	//! get the archive type
+	virtual E_FILE_ARCHIVE_TYPE getType() const { return EFAT_UNKNOWN; }
+};
+
+//! Class which is able to create an archive from a file.
+/** If you want the Irrlicht Engine be able to load archives of
+currently unsupported file formats (e.g .wad), then implement
+this and add your new Archive loader with
+IFileSystem::addArchiveLoader() to the engine. */
+class IArchiveLoader : public virtual core::IReferenceCounted
+{
+public:
+	//! Check if the file might be loaded by this class
+	/** Check based on the file extension (e.g. ".zip")
+	\param filename Name of file to check.
+	\return True if file seems to be loadable. */
+	virtual bool isALoadableFileFormat(const std::filesystem::path& filename) const =0;
+
+	//! Check if the file might be loaded by this class
+	/** This check may look into the file.
+	\param file File handle to check.
+	\return True if file seems to be loadable. */
+	virtual bool isALoadableFileFormat(IReadFile* file) const =0;
+
+	//! Check to see if the loader can create archives of this type.
+	/** Check based on the archive type.
+	\param fileType The archive type to check.
+	\return True if the archile loader supports this type, false if not */
+	virtual bool isALoadableFileFormat(E_FILE_ARCHIVE_TYPE fileType) const =0;
+
+	//! Creates an archive from the file
+	/** \param file File handle to use.
+	\param ignoreCase Searching is performed without regarding the case
+	\param ignorePaths Files are searched for without checking for the directories
+	\return Pointer to newly created archive, or 0 upon error. */
+	virtual core::smart_refctd_ptr<IFileArchive> createArchive(IReadFile* file) const =0;
+};
```

Comment:
We should rethink how this should work.

Reply:
@sadiuk can you find out if its at all possible on any OS to have a file and directory that have the same name in the same path? ( a related question is, can a folder/directory have a `.` in the name)

i.e. lets say we have a directory aptly named
`some/path/archive.zip/simefile.ext`
would it be possible for a standalone file to exist with path
`some/path/archive.zip`

Reply:
@Crisspl @devshgraphicsprogramming on both Windows, Linux, MacOS and Android you cannot have a directory and a file with the same name but can have a directory with `.` in the name.

Reply:
that's nice then we could devise a system to treat archives transparently as directories